File: src/storeFx/fetchTopSongs.ml
open Fx
open MergeIntoMap
open Monad_syntax

let make artistName =
  let* Deps.{ store } = ask () in
  let state = Zustand.getState store in
  let* songs =
    Subsonic.GetTopSongs.make artistName ()
    |. provide ~deps:{ credentials = state.auth.credentials }
  in
  let songs = songs |> Option.value ~default:[||] in
  let topSongIdsByArtistName =
    let state = Zustand.getState store in
    mergeIntoMap state.artists.topSongIdsByArtistName
      [| songs |> Js.Array.map ~f:(fun song -> Subsonic.Song.(song.id)) |]
      (fun _ -> artistName)
  in
  let songsById =
    let state = Zustand.getState store in
    mergeIntoMap state.songs.byId songs (fun song -> song.id)
  in
  if
    let state = Zustand.getState store in
    topSongIdsByArtistName != state.artists.topSongIdsByArtistName
    || songsById != state.songs.byId
  then
    store
    |. Zustand.setState (fun prevState ->
           {
             prevState with
             artists = { prevState.artists with topSongIdsByArtistName };
             songs = { byId = songsById };
           });
  Ok ()

File: src/storeFx/fetchTopSongs.ts
import * as Fx from '../fx/fx';
import { mergeIntoMap } from '../mergeIntoMap/mergeIntoMap';
import type { Store } from '../store/types';
import { getTopSongs } from '../subsonic';

export function fetchTopSongs(artistName: string) {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    return Fx.bind(
      Fx.provide(getTopSongs(artistName), {
        credentials: store.getState().auth.credentials,
      }),
      songs => {
        songs ??= [];

        const state = store.getState();

        const topSongIdsByArtistName = mergeIntoMap(
          state.artists.topSongIdsByArtistName,
          [songs.map(song => song.id)],
          () => artistName,
        );

        const songsById = mergeIntoMap(
          state.songs.byId,
          songs,
          song => song.id,
        );

        if (
          topSongIdsByArtistName !== state.artists.topSongIdsByArtistName ||
          songsById !== state.songs.byId
        ) {
          store.setState(prevState => ({
            artists: { ...prevState.artists, topSongIdsByArtistName },
            songs: { byId: songsById },
          }));
        }

        return Fx.Ok();
      },
    );
  });
}
